app/src/main/java/org/mozilla/fenix/home/blocklist/BlocklistMiddleware.kt

```diff
@@ -0,0 +1,140 @@
+        context: MiddlewareContext<HomeFragmentState, HomeFragmentAction>,
+        next: (HomeFragmentAction) -> Unit,
+        action: HomeFragmentAction
+    ) {
+        val state = context.state
+        next(
+            when (action) {
+                is HomeFragmentAction.Change -> {
+                    action.copy(
+                        recentBookmarks = action.recentBookmarks.filteredByBlocklist(blocklist),
+                        recentTabs = action.recentTabs.filteredByBlocklist(blocklist),
+                    )
+                }
+                is HomeFragmentAction.RecentTabsChange -> {
+                    action.copy(
+                        recentTabs = action.recentTabs.filteredByBlocklist(blocklist)
+                    )
+                }
+                is HomeFragmentAction.RecentBookmarksChange -> {
+                    action.copy(
+                        recentBookmarks = action.recentBookmarks.filteredByBlocklist(blocklist)
+                    )
+                }
+                is HomeFragmentAction.RecentHistoryChange -> {
+                    action.copy(recentHistory = action.recentHistory.filteredByBlocklist(blocklist))
+                }
+                is HomeFragmentAction.RemoveRecentTab -> {
+                    if (action.recentTab is RecentTab.Tab) {
+                        val url = action.recentTab.state.content.url
+                        val updatedBlocklist = blocklist + url
+                        settings.homescreenBlocklist = updatedBlocklist
+                        state.filteredByBlocklist(updatedBlocklist).toChangeAction()
+                    } else {
+                        action
+                    }
+                }
+                is HomeFragmentAction.RemoveRecentBookmark -> {
+                    action.recentBookmark.url?.let {
+                        val updatedBlocklist = blocklist + it
+                        settings.homescreenBlocklist = updatedBlocklist
+                        state.filteredByBlocklist(updatedBlocklist).toChangeAction()
+                    } ?: action
+                }
+                is HomeFragmentAction.RemoveRecentHistoryHighlight -> {
+                    val url = action.highlightUrl
+                    val updatedBlocklist = blocklist + url
+                    settings.homescreenBlocklist = updatedBlocklist
+                    state.filteredByBlocklist(updatedBlocklist).toChangeAction()
+                }
+                else -> action
+            }
+        )
+    }
+
+    private fun HomeFragmentState.toChangeAction() = HomeFragmentAction.Change(
+        topSites = topSites,
+        recentTabs = recentTabs,
+        recentBookmarks = recentBookmarks,
+        recentHistory = recentHistory,
+        mode = mode,
+        collections = collections,
+        tip = tip,
+        showCollectionPlaceholder = showCollectionPlaceholder
+    )
+
+    fun filterState(state: HomeFragmentState): HomeFragmentState =
```

Comment:
This will be required from the linter soon when we upgrade it, but we want to add a kDoc for public functions